File: packages/indexer/src/write-index.ts
import { mkdir, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { gzipSync } from "node:zlib";
import type { Manifest, TermShard } from "@csf/format";
import { encodeTermShardBinary } from "./binary-term-shard.js";
import { contentHash } from "./hash.js";
import type { BuiltIndex } from "./types.js";

/**
 * Recursively sorts object keys (array element order is left alone --
 * it's semantically meaningful for postings/doc-id lists) so the same
 * logical data always serializes to the same bytes regardless of
 * insertion order. `JSON.stringify` alone is deterministic for one
 * producer's own iteration order, but not guaranteed stable across
 * independent producers or a corpus fed in a different order
 * (REVIEW.md#10) -- sorting keys before serializing removes that
 * degree of freedom entirely.
 */
function canonicalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (value !== null && typeof value === "object") {
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      sorted[key] = canonicalize((value as Record<string, unknown>)[key]);
    }
    return sorted;
  }
  return value;
}

async function writeJson(
  outDir: string,
  relPath: string,
  data: unknown,
): Promise<string> {
  const json = JSON.stringify(canonicalize(data));
  const hash = contentHash(json);
  const hashedRelPath = relPath.replace(/\.json$/, `.${hash}.json`);
  const absPath = join(outDir, hashedRelPath);
  await mkdir(dirname(absPath), { recursive: true });
  await writeFile(absPath, json, "utf8");
  return hashedRelPath;
}

async function writeBinary(
  outDir: string,
  relPath: string,
  data: Uint8Array,
): Promise<string> {
  const hash = contentHash(data);
  const hashedRelPath = relPath.replace(/\.bin$/, `.${hash}.bin`);
  const absPath = join(outDir, hashedRelPath);
  await mkdir(dirname(absPath), { recursive: true });
  await writeFile(absPath, data);
  return hashedRelPath;
}

/** Default per docs/02-index-format.md#size-targets--sharding-tuning ("e.g. ~50KB gzipped"). */
export const DEFAULT_MAX_TERM_SHARD_GZIP_BYTES = 50 * 1024;

function groupByPrefixLength(
  termShard: TermShard,
  prefixLength: number,
): Map<string, TermShard> {
  const groups = new Map<string, TermShard>();
  for (const [term, entry] of Object.entries(termShard)) {
    const prefix = term.slice(0, prefixLength);
    let group = groups.get(prefix);
    if (!group) {
      group = {};
      groups.set(prefix, group);
    }
    group[term] = entry;
  }
  return groups;
}

function gzipByteSize(termShard: TermShard): number {
  return gzipSync(JSON.stringify(canonicalize(termShard))).length;
}

/** A prefix length past which auto-splitting gives up even if still over budget -- a safety net against runaway recursion, not a limit expected to bind in practice (docs/02's own examples never go past two characters). */
const MAX_PREFIX_LENGTH = 8;

/**
 * Splits one bucket further by one more prefix character, recursing into
 * any resulting sub-bucket that's *still* over `maxGzipBytes` -- docs/02's
 * "auto-increases prefix length ... for over-large shards" is not capped
 * at two characters: a sufficiently dense leading substring (common in
 * agglutinative/compounding languages like German at real corpus scale)
 * can still be too large even after one split, so this keeps widening
 * per over-large sub-bucket until it fits, terms stop separating further
 * (`subBuckets.size <= 1` -- can't split a bucket by prefix if every term
 * in it already shares that whole prefix), or `MAX_PREFIX_LENGTH` is hit,
 * whichever comes first -- the latter two both emit a warning rather
 * than fail the build, since a corpus dense enough to hit either in
 * practice needs the binary tier this shard format is deliberately
 * staying open-and-simple ahead of (docs/11-binary-vs-json-index.md),
 * not open-ended JSON prefix recursion.
 */
function splitOversizedBucket(
  prefix: string,
  group: TermShard,
  prefixLength: number,
  language: string,
  maxGzipBytes: number,
  result: Map<string, TermShard>,
): void {
  const size = gzipByteSize(group);
  if (size <= maxGzipBytes) {
    result.set(prefix, group);
    return;
  }
  const termCount = Object.keys(group).length;
  if (termCount <= 1 || prefixLength >= MAX_PREFIX_LENGTH) {
    const reason =
      termCount <= 1
        ? "only one term left"
        : `hit the ${MAX_PREFIX_LENGTH}-character prefix-length cap`;
    console.warn(
      `[csf-indexer] term shard [${language}] prefix "${prefix}" is ${size} gzip bytes, over the ${maxGzipBytes}-byte budget and cannot be split further (${reason}) -- see docs/02-index-format.md#size-targets--sharding-tuning.`,
    );
    result.set(prefix, group);
    return;
  }
  const subBuckets = groupByPrefixLength(group, prefixLength + 1);
  if (subBuckets.size <= 1) {
    console.warn(
      `[csf-indexer] term shard [${language}] prefix "${prefix}" is ${size} gzip bytes, over the ${maxGzipBytes}-byte budget, but every term in it shares the same prefix so it cannot be split further -- see docs/02-index-format.md#size-targets--sharding-tuning.`,
    );
    result.set(prefix, group);
    return;
  }
  for (const [subPrefix, subGroup] of subBuckets) {
    splitOversizedBucket(
      subPrefix,
      subGroup,
      prefixLength + 1,
      language,
      maxGzipBytes,
      result,
    );
  }
}

/**
 * Splits one language's term shard into prefix buckets
 * (docs/02-index-format.md#term-shard-inverted-index): single-character
 * prefix by default, so a query for "widget" only ever needs the "w"
 * bucket, regardless of how many other terms exist. A bucket that would
 * still exceed `maxGzipBytes` after gzip is widened one character at a
 * time (`splitOversizedBucket` above) rather than left as one oversized
 * fetch every one-character-prefix query would have to pay for.
 */
function shardTermsByPrefix(
  termShard: TermShard,
  language: string,
  maxGzipBytes: number,
): Map<string, TermShard> {
  const result = new Map<string, TermShard>();
  for (const [prefix, group] of groupByPrefixLength(termShard, 1)) {
    splitOversizedBucket(prefix, group, 1, language, maxGzipBytes, result);
  }
  return result;
}

/**
 * Serializes a BuiltIndex to content-hashed shard files plus a plain
 * (unhashed) manifest.json — the full hashed-manifest + alias-pointer
 * scheme from docs/02-index-format.md#versioning--cache-strategy is a
 * deliberate simplification left for when a real deployment needs it;
 * shard immutability (the property that actually matters for caching)
 * is implemented as designed.
 *
 * Builds a fresh manifest object rather than mutating `built.manifest`
 * in place, so the `BuiltIndex` a caller got back from `buildIndex()`
 * stays exactly what it was — a caller holding onto that reference
 * shouldn't see it silently change shape out from under them just
 * because they also called `writeIndex()` (REVIEW.md#9).
 */
export interface WriteIndexOptions {
  /** See `DEFAULT_MAX_TERM_SHARD_GZIP_BYTES` and `shardTermsByPrefix()` above. */
  maxShardGzipBytes?: number;
  /**
   * `false` writes exactly one term shard per language (`prefix: "all"`)
   * instead of splitting by first-character prefix -- the same shard
   * *format* either way, just configured with one shard instead of many.
   * Real per-prefix sharding is solving a fetch-size problem that
   * doesn't exist yet for a small enough corpus
   * (docs/14-reference-deployment-cms-2k.md#what-to-simplify-at-this-scale).
   * Defaults to `true` (real prefix sharding, the general-case default).
   */
  shardByPrefix?: boolean;
  /**
   * `"binary"` writes term shards with the directory-based delta+varint
   * encoding (`./binary-term-shard.js`, validated in
   * `packages/indexer/bench/binary-lazy-decode.mjs` — see
   * docs/11-binary-vs-json-index.md), one `.bin` file per prefix bucket
   * with `format: "binary"` recorded on that shard's manifest entry, in
   * place of the plain-JSON `terms/<lang>/<prefix>.json` shape.
   * Defaults to `"json"`. Every other shard type (facets, doc store,
   * pins, synonyms, fuzzy) stays JSON either way — this option is
   * term-shard-only, matching docs/spec-binary-format.md's "a
   * deployment may mix JSON and binary files" allowance.
   */
  termShardFormat?: "json" | "binary";
}

export async function writeIndex(
  built: BuiltIndex,
  outDir: string,
  options: WriteIndexOptions = {},
): Promise<void> {
  const maxGzipBytes =
    options.maxShardGzipBytes ?? DEFAULT_MAX_TERM_SHARD_GZIP_BYTES;
  const shardByPrefix = options.shardByPrefix ?? true;
  const termShardFormat = options.termShardFormat ?? "json";
  const languages = Object.keys(built.termShards).sort();
  const terms = (
    await Promise.all(
      languages.map(async (language) => {
        const termShard = built.termShards[language] ?? {};
        const buckets = shardByPrefix
          ? [...shardTermsByPrefix(termShard, language, maxGzipBytes)].sort(
              ([a], [b]) => a.localeCompare(b),
            )
          : ([["all", termShard]] as [string, TermShard][]);
        return Promise.all(
          buckets.map(async ([prefix, group]) => {
            const file =
              termShardFormat === "binary"
                ? await writeBinary(
                    outDir,
                    `terms/${language}/${prefix}.bin`,
                    encodeTermShardBinary(group),
                  )
                : await writeJson(
                    outDir,
                    `terms/${language}/${prefix}.json`,
                    group,
                  );
            return {
              lang: language,
              prefix,
              file,
              termCount: Object.keys(group).length,
              ...(termShardFormat === "binary"
                ? { format: "binary" as const }
                : {}),
            };
          }),
        );
      }),
    )
  ).flat();

  const docsFile = await writeJson(outDir, "docs/0.json", built.docStore);
  const docs = [{ shard: 0, file: docsFile, idRange: built.idRange }];

  const facetFields = Object.keys(built.facetShards).sort();
  const facets = facetFields.length
    ? await Promise.all(
        facetFields.map(async (field) => ({
          field,
          file: await writeJson(
            outDir,
            `facets/${field}.json`,
            built.facetShards[field],
          ),
        })),
      )
    : undefined;

  const pinLanguages = Object.keys(built.pinsShards)
    .filter((language) => Object.keys(built.pinsShards[language] ?? {}).length)
    .sort();
  let pins: Record<string, string> | undefined;
  if (pinLanguages.length) {
    pins = {};
    for (const language of pinLanguages) {
      pins[language] = await writeJson(
        outDir,
        `pins/${language}.json`,
        built.pinsShards[language],
      );
    }
  }

  const synonymLanguages = Object.keys(built.synonymShards)
    .filter((language) => {
      const shard = built.synonymShards[language];
      return (
        (shard?.equivalences?.length ?? 0) > 0 ||
        Object.keys(shard?.directional ?? {}).length > 0 ||
        (shard?.multiWord?.length ?? 0) > 0
      );
    })
    .sort();
  let synonyms: Record<string, string> | undefined;
  if (synonymLanguages.length) {
    synonyms = {};
    for (const language of synonymLanguages) {
      synonyms[language] = await writeJson(
        outDir,
        `synonyms/${language}.json`,
        built.synonymShards[language],
      );
    }
  }

  const fuzzyLanguages = Object.keys(built.fuzzyShards)
    .filter(
      (language) =>
        Object.keys(built.fuzzyShards[language]?.deletions ?? {}).length,
    )
    .sort();
  let fuzzy: Record<string, string> | undefined;
  if (fuzzyLanguages.length) {
    fuzzy = {};
    for (const language of fuzzyLanguages) {
      fuzzy[language] = await writeJson(
        outDir,
        `fuzzy/${language}.json`,
        built.fuzzyShards[language],
      );
    }
  }

  const manifest: Manifest = {
    ...built.manifest,
    shards: {
      terms,
      docs,
      ...(facets ? { facets } : {}),
    },
    ...(pins ? { pins } : {}),
    ...(synonyms ? { synonyms } : {}),
    ...(fuzzy ? { fuzzy } : {}),
  };

  await mkdir(outDir, { recursive: true });
  await writeFile(
    join(outDir, "manifest.json"),
    JSON.stringify(canonicalize(manifest)),
    "utf8",
  );
}

File: python/searchable/src/searchable/indexer/write_index.py
import gzip
import json
import re
from pathlib import Path
from typing import Any

from searchable.indexer.hash import content_hash
from searchable.indexer.types import BuiltIndex

DEFAULT_MAX_TERM_SHARD_GZIP_BYTES = 50 * 1024
_MAX_PREFIX_LENGTH = 8


def _canonicalize(value: object) -> object:
    if isinstance(value, list):
        return [_canonicalize(v) for v in value]
    if isinstance(value, tuple):
        return [_canonicalize(v) for v in value]
    if isinstance(value, dict):
        return {key: _canonicalize(value[key]) for key in sorted(value.keys())}
    return value


def _to_json(data: object) -> str:
    return json.dumps(_canonicalize(data), separators=(",", ":"), ensure_ascii=False)


def _write_json(out_dir: str, rel_path: str, data: object) -> str:
    content = _to_json(data)
    digest = content_hash(content)
    hashed_rel_path = re.sub(r"\.json$", f".{digest}.json", rel_path)
    abs_path = Path(out_dir) / hashed_rel_path
    abs_path.parent.mkdir(parents=True, exist_ok=True)
    abs_path.write_text(content, encoding="utf-8")
    return hashed_rel_path


def _gzip_byte_size(term_shard: dict[str, Any]) -> int:
    return len(gzip.compress(_to_json(term_shard).encode("utf-8")))


def _group_by_prefix_length(
    term_shard: dict[str, Any], prefix_length: int
) -> dict[str, dict[str, Any]]:
    groups: dict[str, dict[str, Any]] = {}
    for term, entry in term_shard.items():
        prefix = term[:prefix_length]
        groups.setdefault(prefix, {})[term] = entry
    return groups


def _split_oversized_bucket(
    prefix: str,
    group: dict[str, Any],
    prefix_length: int,
    language: str,
    max_gzip_bytes: int,
    result: dict[str, dict[str, Any]],
) -> None:
    size = _gzip_byte_size(group)
    if size <= max_gzip_bytes:
        result[prefix] = group
        return
    term_count = len(group)
    if term_count <= 1 or prefix_length >= _MAX_PREFIX_LENGTH:
        result[prefix] = group
        return
    sub_buckets = _group_by_prefix_length(group, prefix_length + 1)
    if len(sub_buckets) <= 1:
        result[prefix] = group
        return
    for sub_prefix, sub_group in sub_buckets.items():
        _split_oversized_bucket(
            sub_prefix, sub_group, prefix_length + 1, language, max_gzip_bytes, result
        )


def _shard_terms_by_prefix(
    term_shard: dict[str, Any], language: str, max_gzip_bytes: int
) -> dict[str, dict[str, Any]]:
    result: dict[str, dict[str, Any]] = {}
    for prefix, group in _group_by_prefix_length(term_shard, 1).items():
        _split_oversized_bucket(prefix, group, 1, language, max_gzip_bytes, result)
    return result


def _chunk_doc_store_by_id_range(
    doc_store: dict[str, Any], shard_size: float
) -> list[dict[str, Any]]:
    sorted_ids = sorted(int(k) for k in doc_store.keys())
    chunks: list[dict[str, Any]] = []
    step = len(sorted_ids) if shard_size == float("inf") else int(shard_size)
    step = max(step, 1)
    i = 0
    while i < len(sorted_ids):
        ids_in_chunk = sorted_ids[i : i + step]
        shard = {str(doc_id): doc_store[str(doc_id)] for doc_id in ids_in_chunk}
        chunks.append({"idRange": (ids_in_chunk[0], ids_in_chunk[-1]), "shard": shard})
        i += step
    return chunks


def write_index(
    built: BuiltIndex,
    out_dir: str,
    max_shard_gzip_bytes: int = DEFAULT_MAX_TERM_SHARD_GZIP_BYTES,
    shard_by_prefix: bool = True,
    doc_store_shard_size: float = float("inf"),
) -> None:
    languages = sorted(built.term_shards.keys())
    terms: list[dict[str, Any]] = []
    for language in languages:
        term_shard = built.term_shards.get(language, {})
        if shard_by_prefix:
            buckets = sorted(
                _shard_terms_by_prefix(term_shard, language, max_shard_gzip_bytes).items()
            )
        else:
            buckets = [("all", term_shard)]
        for prefix, group in buckets:
            file = _write_json(out_dir, f"terms/{language}/{prefix}.json", group)
            terms.append(
                {
                    "lang": language,
                    "prefix": prefix,
                    "file": file,
                    "termCount": len(group),
                }
            )

    doc_store_chunks = _chunk_doc_store_by_id_range(built.doc_store, doc_store_shard_size)
    if not doc_store_chunks:
        doc_store_chunks = [{"idRange": built.id_range, "shard": {}}]
    docs: list[dict[str, Any]] = []
    for shard_index, chunk in enumerate(doc_store_chunks):
        file = _write_json(out_dir, f"docs/{shard_index}.json", chunk["shard"])
        docs.append({"shard": shard_index, "file": file, "idRange": list(chunk["idRange"])})

    facet_fields = sorted(built.facet_shards.keys())
    facets = None
    if facet_fields:
        facets = []
        for field_name in facet_fields:
            file = _write_json(out_dir, f"facets/{field_name}.json", built.facet_shards[field_name])
            facets.append({"field": field_name, "file": file})

    pin_languages = sorted(language for language, shard in built.pins_shards.items() if shard)
    pins = None
    if pin_languages:
        pins = {}
        for language in pin_languages:
            pins[language] = _write_json(
                out_dir, f"pins/{language}.json", built.pins_shards[language]
            )

    def _synonym_shard_nonempty(shard: dict[str, Any]) -> bool:
        return bool(shard.get("equivalences") or shard.get("directional") or shard.get("multiWord"))

    synonym_languages = sorted(
        language
        for language, shard in built.synonym_shards.items()
        if _synonym_shard_nonempty(shard)
    )
    synonyms = None
    if synonym_languages:
        synonyms = {}
        for language in synonym_languages:
            synonyms[language] = _write_json(
                out_dir, f"synonyms/{language}.json", built.synonym_shards[language]
            )

    fuzzy_languages = sorted(
        language for language, shard in built.fuzzy_shards.items() if shard.get("deletions")
    )
    fuzzy = None
    if fuzzy_languages:
        fuzzy = {}
        for language in fuzzy_languages:
            file = _write_json(out_dir, f"fuzzy/{language}.json", built.fuzzy_shards[language])
            fuzzy[language] = {"file": file}

    manifest = {
        **built.manifest,
        "shards": {
            "terms": terms,
            "docs": docs,
            **({"facets": facets} if facets else {}),
        },
        **({"pins": pins} if pins else {}),
        **({"synonyms": synonyms} if synonyms else {}),
        **({"fuzzy": fuzzy} if fuzzy else {}),
    }
    manifest.pop("format", None)

    out_path = Path(out_dir)
    out_path.mkdir(parents=True, exist_ok=True)
    (out_path / "manifest.json").write_text(_to_json(manifest), encoding="utf-8")
